File: projeto_brademo/lib/service/listaService.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/apiConfig.dart';
import '../model/lista.dart';

class ListaService {
  Future<Lista?> criarLista(int idUsuario, String nomeLista) async {
    try {
      final response = await http.post(
        Uri.parse('${ApiConfig.baseUrl}/listas/$idUsuario'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'nome_lista': nomeLista}),
      );
      if (response.statusCode == 200 || response.statusCode == 201) {
        return Lista.fromJson(jsonDecode(response.body));
      }

      return null;
    } catch (e) {
      return null;
    }
  }

  Future<bool> adicionarFilme(int idLista, int idFilme) async {
    try {
      final response = await http.post(
        Uri.parse('${ApiConfig.baseUrl}/listas/$idLista/filme/$idFilme'),
      );

      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  Future<List<Lista>> buscarListasUsuario(int idUsuario) async {
    try {
      final response = await http.get(
        Uri.parse('${ApiConfig.baseUrl}/listas/usuario/$idUsuario'),
      );

      if (response.statusCode == 200) {
        final List<dynamic> dados = jsonDecode(response.body);
        return dados.map((lista) => Lista.fromJson(lista)).toList();
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  Future<List<Lista>> buscarListasDoFilme(int idUsuario, int idFilme) async {
    try {
      final response = await http.get(
        Uri.parse(
          '${ApiConfig.baseUrl}/listas/filme/$idFilme/usuario/$idUsuario',
        ),
      );
      if (response.statusCode == 200) {
        final List<dynamic> dados = jsonDecode(response.body);
        return dados.map((lista) => Lista.fromJson(lista)).toList();
      }
      return [];
    } catch (e) {
      return [];
    }
  }

  Future<bool> removerFilme(int idLista, int idFilme) async {
    try {
      final response = await http.delete(
        Uri.parse('${ApiConfig.baseUrl}/listas/$idLista/filme/$idFilme'),
      );
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }

  Future<Lista?> editarLista(int idLista, String nomeLista) async {
    try {
      final response = await http.put(
        Uri.parse('${ApiConfig.baseUrl}/listas/$idLista'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({'nome_lista': nomeLista}),
      );

      if (response.statusCode == 200) {
        return Lista.fromJson(jsonDecode(response.body));
      }
      return null;
    } catch (e) {
      return null;
    }
  }

  Future<bool> excluirLista(int idLista) async {
    try {
      final response = await http.delete(
        Uri.parse('${ApiConfig.baseUrl}/listas/$idLista'),
      );

      return response.statusCode == 200 || response.statusCode == 204;
    } catch (e) {
      return false; 
    }
  }
}

File: moviematch_back/src/main/java/com/brademobackend/moviematch/service/ListaService.java
package com.brademobackend.moviematch.service;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.brademobackend.moviematch.model.Filme;
import com.brademobackend.moviematch.model.Lista;
import com.brademobackend.moviematch.model.Usuario;
import com.brademobackend.moviematch.repository.FilmeRepository;
import com.brademobackend.moviematch.repository.ListaRepository;
import com.brademobackend.moviematch.repository.UsuarioRepository;

@Service
public class ListaService {

    @Autowired
    private ListaRepository listaRepository;
    @Autowired
    private UsuarioRepository usuarioRepository;
    @Autowired
    private FilmeRepository filmeRepository;

    public Lista criarLista(Long idUsuario, String nomeLista) {
        Usuario usuario = usuarioRepository.findById(idUsuario).orElseThrow();
        Lista lista = new Lista();
        lista.setNome_lista(nomeLista);
        lista.setUsuario(usuario);
        return listaRepository.save(lista);
    }

    public Lista adicionarFilme(Long idLista, Long idFilme) {
        Lista lista = listaRepository.findById(idLista).orElseThrow();
        Filme filme = filmeRepository.findById(idFilme).orElseThrow();

        boolean jaExiste = lista.getFilmes().stream().anyMatch(f -> f.getId_filme().equals(idFilme));
        if (!jaExiste) {
            lista.getFilmes().add(filme);
        }
        return listaRepository.save(lista);
    }

    public Iterable<Lista> listarListas() {
        return listaRepository.findAll();
    }

    public List<Lista> buscarListasUsuario(Long idUsuario) {
        return listaRepository.buscarPorUsuario(idUsuario);
    }

    public List<Lista> buscarListasDoFilme(Long idUsuario, Long idFilme) {
        return listaRepository.buscarListasDoFilme(idUsuario, idFilme);
    }

    public Lista removerFilme(Long idLista, Long idFilme) {
        Lista lista = listaRepository.findById(idLista).orElseThrow();
        lista.getFilmes().removeIf(filme -> filme.getId_filme().equals(idFilme));
        return listaRepository.save(lista);
    }

    public Lista editarLista(Long idLista, String nomeLista) {
        Lista lista = listaRepository.findById(idLista).orElseThrow();
        lista.setNome_lista(nomeLista);
        return listaRepository.save(lista);
    }

    public void excluirLista(Long idLista) {
        listaRepository.deleteById(idLista);
    }
}